src/safeposix/syscalls/fs_calls.rs

```
@@ -781,47 +847,69 @@ impl Cage {
                 let retval = match metawalkandparent(truenewpath.as_path()) {
+                    // If both the file and the parent doesn't exist, newpath can't be created
                     (None, None) => {
                         syscall_error(Errno::ENOENT, "link", "newpath cannot be created")
                     }
 
+                    // If the newpath exists, linking can't be perfomed and an error is returned.
+                    (Some(_), ..) => syscall_error(Errno::EEXIST, "link", "newpath already exists"),
+
+                    // If the parent directory inode exists, make a reference of the oldpath inode
+                    // in the parent directory to make a link between the two directory paths.
                     (None, Some(pardirinode)) => {
+                        // Get the mutable instance of the parent inode object
                         let mut parentinodeobj =
                             FS_METADATA.inodetable.get_mut(&pardirinode).unwrap();
                         //insert a reference to the inode in the parent directory
                         if let Inode::Dir(ref mut parentdirinodeobj) = *parentinodeobj {
                             parentdirinodeobj
                                 .filename_to_inode_dict
                                 .insert(filename, inodenum);
+                            // Increment the link count of the parent inode as well because
+                            // when a link is created, a new directory entry is added to
+                            // the parent directory of the new link. 
                             parentdirinodeobj.linkcount += 1;
+                            //drop the mutable instance of the parent inode object
                             drop(parentinodeobj);
                             log_metadata(&FS_METADATA, pardirinode);
                             log_metadata(&FS_METADATA, inodenum);
                         } else {
```

Why is this supposed to be untriggerable by the user?  Why panic instead of error?

---

Can you please explain the first part of the question: "Why is this supposed to be untriggerable by the user?". 
For the second question, I referenced a couple of other areas where this same code snippet is used, somewhere we have returned `Panics` and other areas `Errors`, which I think should be consistent everwhere. 
I thought of leaving it as it is.

---

If you use panic, it should only be for a situation where there is some internal state that is messed up and you can't recover.   In other words, a user giving bad input should not be able to get the system into such a state.  This is what I was referring to.

---

Understood, then its better to throw an error in such cases. Will update it to return an error.